### PR TITLE
MET EE noise mitigation in 2017 data

### PR DESCRIPTION
**Pull request by eioannou**
This Pull Request includes the recipe which was created for the EE noise mitigation in 2017 data. It removes bad pf candidates clustered to jets by adding new modules for fix EE in MET tool. It adds the fixEE2017 and fixEE2017Params parameters to MET tool.

This recipe creates a new pfcandidates collection which does not include the lower pt jets in the high-eta endcap region which are mostly affected by EE noise. It removes all jets and associated PF candidates with pt<50 GeV and 2.65 <|eta| < 3.139 before calculating MET.

More details of the recipe can be found here: https://indico.cern.ch/event/759372/contributions/3149378/attachments/1721436/2802416/metreport.pdf

cc: @peruzzim

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: JetMETCorrections/Type1MET/plugins/BadPFCandidateJetsEEnoiseProducer.cc [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+namespace pat {
+  class BadPFCandidateJetsEEnoiseProducer : public edm::global::EDProducer<>{
+  public:
+    explicit BadPFCandidateJetsEEnoiseProducer(const edm::ParameterSet&);
+    ~BadPFCandidateJetsEEnoiseProducer() override;
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+    void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+  private:
+    edm::EDGetTokenT<edm::View<pat::Jet> > jetsrc_;
+    double PtThreshold_;
```
Comment (perrotta) [inline JetMETCorrections/Type1MET/plugins/BadPFCandidateJetsEEnoiseProducer.cc]: Use lower case initials for the class data members (CMS standard)

File: JetMETCorrections/Type1MET/plugins/BadPFCandidateJetsEEnoiseProducer.cc [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    double MinEtaThreshold_;
+    double MaxEtaThreshold_;
+    bool userawPt_;
+  };
+}
+
+
+pat::BadPFCandidateJetsEEnoiseProducer::BadPFCandidateJetsEEnoiseProducer(const edm::ParameterSet& iConfig) :
+  jetsrc_(consumes<edm::View<pat::Jet> >(iConfig.getParameter<edm::InputTag>("jetsrc") )),
+  PtThreshold_(iConfig.getParameter<double>("PtThreshold")),
```
Comment (perrotta) [inline JetMETCorrections/Type1MET/plugins/BadPFCandidateJetsEEnoiseProducer.cc]: Lower case initials also for the configuration parameters

File: JetMETCorrections/Type1MET/plugins/BadPFCandidateJetsEEnoiseProducer.cc [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+ 
+  // find the bad jets
+  for (int jetindex = 0; jetindex < njets; ++jetindex){
+    edm::Ptr<pat::Jet> candjet = jetcandidates->ptrAt(jetindex);
+
+    // Corrected Pt or Uncorrected Pt (It is defined from cfi file)
+    double PtJet = userawPt_ ? candjet->correctedJet("Uncorrected").pt() : candjet->pt();
+    double AbsEtaJet = std::abs(candjet->eta());
+    
+    if ( PtJet > PtThreshold_ || AbsEtaJet < MinEtaThreshold_ || AbsEtaJet > MaxEtaThreshold_) {
```
Comment (perrotta) [inline JetMETCorrections/Type1MET/plugins/BadPFCandidateJetsEEnoiseProducer.cc]: Lower than Min or larger than Max? Please check

Comment (perrotta) [reply JetMETCorrections/Type1MET/plugins/BadPFCandidateJetsEEnoiseProducer.cc]: Assuming this is correct, please explain the logic in the PR description, and/or link some slide that describes it

**Comment by perrotta** [comment]
Is this expected to be run on data, or just for private analyses.
If it runs in some data workflow, what should we expect to see in output (nothing, according to the results of the jenkins tests...)

**Comment by perrotta** [comment]
What is the plan for this PR? It is just meant to 10_4_X (i.e. the release for the super-legacy reprocessing), or do you plan to backport it into some analysis release?

**Comment by eioannou** [comment]
We plan to backport it into CMSSW_10_2_X and CMSSW_9_4_X as it was requested by NanoAOD team (@peruzzim). We want to add the MET v2 recipe for all 2017 nanoAOD.

**Comment by slava77** [comment]
@perrotta 
please note that the validate reco comparisons are broken now for miniAOD files https://github.com/cms-sw/cmsdist/pull/4417#issuecomment-430541660
so, if any changes are expected in miniAOD from this PR, perhaps we better wait until ROOT distribution is fixed before signing

**Comment by peruzzim** [comment]
I confirm we would need this functionality backported to 9_4_X and 10_2_X releases, to run it on top of miniAOD in nanoAOD and private ntuple productions.
We will also consider adding it to miniAOD, if a re-miniAOD of 2017 is run before the ultra-legacy re-reco.

**Comment by slava77** [comment]
On 10/18/18 12:07 AM, peruzzim wrote:
> I confirm we would need this functionality backported to 9_4_X and
> 10_2_X releases, to run it on top of miniAOD in nanoAOD and private
> ntuple productions.
> We will also consider adding it to miniAOD, if a re-miniAOD of 2017 is
> run before the ultra-legacy re-reco.
> 
> —

OK.
Thanks for the clarification.
The backports will have to satisfy the policy of not changing the
default behavior.

**Comment by slava77** [comment]
@cmsbuild please test

to get cleaner test results
(as well as also having working fwlite-based reco comparisons)

File: JetMETCorrections/Type1MET/plugins/BadPFCandidateJetsEEnoiseProducer.cc [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  iEvent.put(std::move(goodJets),"good");
+  iEvent.put(std::move(badJets),"bad");
+  
+}
+
+void pat::BadPFCandidateJetsEEnoiseProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("jetsrc",edm::InputTag("slimmedJets"));
+  desc.add<bool>("userawPt",true);
+  desc.add<double>("ptThreshold",75.0);
```
Comment (perrotta) [inline JetMETCorrections/Type1MET/plugins/BadPFCandidateJetsEEnoiseProducer.cc]: Why the default threshold is 75 GeV (also in cleanJetFromEEnoise_cfi.py), and it is only modified for 'fixEE2017Params'?

**Comment by perrotta** [comment]
+1
- By using the default parameters in ```PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py``` the newly added filter is disabled, and therefore no effect can be propagated to the standard workflows (and no effect is seen in the results of the jenkins tests, in fact): as such the pull request can be merged without modifying the default behaviour, and it is therefore suitable for being backported.
- By setting by hand the flag to true I could verify the effect in the miniAOD output on a TTbar sample (it was wf 11624, TTbar MC in 2019 configuration: the one that I had at hand). For example:
all_mini_testpr24876vsorig_ttbar13tev2019wf11624p0c_patmets_slimmedmetspuppi__reco_obj_pt.
  - The flag can be switched on in analyses or in some possible dedicated re-miniAOD campaign

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Pull request merged**